mlir/lib/Dialect/Rock/Generator/Conv2dGenerator.cpp

```diff
@@ -875,14 +883,41 @@ LogicalResult Conv2dGenerator::genConvModule(ModuleOp &module, int kernel_id,
     block->push_front(convOp);
   } break;
   case ConvOpType::BwdData: {
-    auto convOp = builder.create<Conv2DBwdDataOp>(
-        builder.getUnknownLoc(), ArrayRef<Type>{}, args, attributes);
-    block->push_front(convOp);
+    if (kernelId < 0) {
+      // zero init input tensor
+      auto zeroInit = builder.create<ZeroInitKernelOp>(
+          builder.getUnknownLoc(), /*resultType=*/TypeRange{}, args[1],
+          archStrAttr, /*blockSize=*/nullptr, /*gridSize=*/nullptr,
+          /*elemsPerThread=*/nullptr);
+      block->push_front(zeroInit);
+    } else {
+      auto convOp = builder.create<Conv2DBwdDataOp>(
+          builder.getUnknownLoc(), ArrayRef<Type>{}, args, attributes);
+      block->push_front(convOp);
+    }
   } break;
   case ConvOpType::BwdWeight: {
-    auto convOp = builder.create<Conv2DBwdWeightOp>(
-        builder.getUnknownLoc(), ArrayRef<Type>{}, args, attributes);
-    block->push_back(convOp);
+    bool hasUtilities = getBwdWeightKernelCount(builder) > 1;
+    if (hasUtilities && kernelId == 0) {
+      // If there is a workspace, zero-init it, otherwise fill the filter tensor
+      auto zeroInitOp = builder.create<ZeroInitKernelOp>(
```

Comment:
It is nice cleanup to move this chunk of logic into conv2d generator where it should belong. Thanks!